halo2_proofs/src/dev.rs

```diff
@@ -565,354 +564,7 @@ impl<F: FieldExt> MockProver<F> {
-        let perm_errors = {
-            // Original values of columns involved in the permutation.
-            let original = |column, row| {
-                self.cs
-                    .permutation
-                    .get_columns()
-                    .get(column)
-                    .map(|c: &Column<Any>| match c.column_type() {
-                        Any::Advice => self.advice[c.index()][row],
-                        Any::Fixed => self.fixed[c.index()][row],
-                        Any::Instance => CellValue::Assigned(self.instance[c.index()][row]),
-                    })
-                    .unwrap()
-            };
-
-            // Iterate over each column of the permutation
-            self.permutation
-                .mapping
-                .iter()
-                .enumerate()
-                .flat_map(move |(column, values)| {
-                    // Iterate over each row of the column to check that the cell's
-                    // value is preserved by the mapping.
-                    values.iter().enumerate().filter_map(move |(row, cell)| {
-                        let original_cell = original(column, row);
-                        let permuted_cell = original(cell.0, cell.1);
-                        if original_cell == permuted_cell {
-                            None
-                        } else {
-                            let columns = self.cs.permutation.get_columns();
-                            let column = columns.get(column).unwrap();
-                            Some(VerifyFailure::Permutation {
-                                column: (*column).into(),
-                                location: FailureLocation::find(
-                                    &self.regions,
-                                    row,
-                                    Some(column).into_iter().cloned().collect(),
-                                ),
-                            })
-                        }
-                    })
-                })
-        };
-
-        let mut errors: Vec<_> = iter::empty()
-            .chain(selector_errors)
-            .chain(gate_errors)
-            .chain(lookup_errors)
-            .chain(perm_errors)
-            .collect();
-        if errors.is_empty() {
-            Ok(())
-        } else {
-            // Remove any duplicate `ConstraintPoisoned` errors (we check all unavailable
-            // rows in case the trigger is row-specific, but the error message only points
-            // at the constraint).
-            errors.dedup_by(|a, b| match (a, b) {
-                (
-                    a @ VerifyFailure::ConstraintPoisoned { .. },
-                    b @ VerifyFailure::ConstraintPoisoned { .. },
-                ) => a == b,
-                _ => false,
-            });
-            Err(errors)
-        }
+        self.verify_at_rows_par(self.usable_rows.clone(), self.usable_rows.clone())
```

Comment:
it works. But sometimes we deliberately use single thread in proving. (eg: cargo test launch tests parallelly so it bring no benefit (if not worse) to use multi thread innerly.

Reply:
right. Seems like in the PSE's integration, they have removed `self.verify_at_rows(...)` and the only option here is `self.verify_at_rows_par`